src/xtgeo/grid3d/_gridprop_import_xtgcpprop.py

```diff
@@ -1,18 +1,67 @@
 """GridProperty import function of xtgcpprop format."""
+from __future__ import annotations
 
 import json
+from collections.abc import Sequence
+from contextlib import contextmanager
+from io import BytesIO, StringIO
+from pathlib import Path
 from struct import unpack
+from typing import TYPE_CHECKING, Any, Generator
 
 import numpy as np
 
-import xtgeo
 import xtgeo.common.sys as xsys
 from xtgeo.common import null_logger
+from xtgeo.common.constants import UNDEF, UNDEF_INT
+from xtgeo.metadata.metadata import MetaDataCPProperty
 
 logger = null_logger(__name__)
 
+if TYPE_CHECKING:
+    from numpy.typing import DTypeLike
 
-def import_xtgcpprop(mfile, ijrange=None, zerobased=False):
+    from xtgeo.common.sys import _XTGeoFile
+
+
+@contextmanager
+def _read_filelike(
+    filelike: Path | BytesIO | StringIO,
+    offset: int,
+) -> Generator[bytes, None, None]:
+    """Context manager for reading a specified number of bytes from a file-like object.
+
+    Accepts either a Path, BytesIO, or StringIO object as input. The function reads
+    up to 'offset' bytes from the file-like object and yields these bytes.
+
+    For BytesIO and StringIO, the read operation preserves the original
+    file cursor position.
+    """
+
+    if isinstance(filelike, Path):
+        with filelike.open("rb") as f:
+            yield f.read(offset)
+    elif isinstance(filelike, BytesIO):
+        was_at = filelike.tell()
+        try:
+            yield filelike.read(offset)
+        finally:
+            filelike.seek(was_at)
+    elif isinstance(filelike, StringIO):
+        was_at = filelike.tell()
+        try:
+            yield filelike.read(offset).encode()
+        finally:
+            filelike.seek(was_at)
```

Comment:
Could be combined into one-ish, but i like this more. I tried and it just looks like a mess.